books/2015/pro-android-web-game-apps-2012/summary/Chapter5.md

```diff
@@ -463,45 +462,45 @@ give us the image the user clicked on.
 Drawing a canvas on top of another one is a simple call to `drawImage`, which
 happens to accept a canvas as its first argument.
 
-As an untested sketch:
-
-```javascript
-var shapes = [{image: "...", x: 14, y: 18}, /* ... */];
-var canvas = document.getElementById("visibleCanvas");
-var buffer = document.createElement("canvas");
-var masks = document.createElement("canvas");
-
-var width = masks.width = canvas.width;
-var height = masks.height = canvas.height;
-
-function color_to_index(color) {
-  /* ... */
-}
-
-function index_to_color(index) {
-  /* ... */
-}
-
-function get_shape_index(x, y) {
-  var m = masks.getContext('2d');
-  m.clearRect(0, 0, width, height);
-  shapes.forEach((shape, index) => {
-    draw_mask(shape.image, shape.x, shape.y, index_to_color(index), m);
-  });
-  return color_to_index(m.getImageData(x, y, 1, 1).data);
-}
-
-function draw_mask(source, x, y, color, target) {
-  var w = buffer.width = source.width;
-  var h = buffer.height = source.height;
-  var b = buffer.getContext('2d');
-  b.drawImage(source, 0, 0, w, h, 0, 0, w, h);
-  b.globalCompositeOperation = 'source-atop';
-  b.fillStyle = color;
-  b.fillRect(0, 0, w, h);
-  target.drawImage(buffer, x, y);
-}
-```
+> As an untested sketch:
```

Comment:
I don’t understand the need for both mask and buffer

Reply:
🤔

Assuming the following explanation helps, please suggest edits to the text that would make it clearer.

We have one canvas that is drawn on the screen. That is the one the user sees. Let's call it `drawn`.

We want another canvas, which I call `masks`, which represents the same set of objects, drawn in the same order (if they overlap, this is important), at the same positions, but where each object being drawn instead of being a full-colored sprite, is a single solid color.

If each object has a different solid color, and we have painted the two canvasses in the same way except for the source images (full-colors vs. one color), we can look at [the color of [the pixel coordinates the user clicked on `drawn`] on `masks`] to figure out which object has been clicked.

So, two canvasses.

But we don't have the masks. One option would be to ask our graphic artist to provide them, or we could somehow pregenerate them manually and store a second sprite sheet with the masks for each object. What we do here, though, is that we want to generate the masks on the fly. There is a reasonably easy way to do that with `source-atop`, but that requires a canvas on which we:
- draw the original sprite
- change the composition operation
- draw a monochrome rectangle
This canvas cannot be `drawn`, as we don't want the user to see it, and it cannot be `masks`, because then the second sprite we try to add would color over the first one and we would not have one color per sprite anymore. The simplest approach here would be to have a canvas per sprite dedicated to creating a mask for that one sprite.

But we only need that canvas for the brief time it takes to create the monochrome sprite (= mask) and copy it over to `masks`. After that, we move on to the next image, and we want a blank canvas again to create the next mask.

The book seems to assume that creating a new `canvas` element is more expensive than resizing and clearing an existing one. I have not verified that, but it is a sentiment I have seen expressed or implied many times over the past twenty years.

So we create a single canvas, `buffer`, for the purpose of creating individual masks one at a time and copying them to the `masks` canvas.

Among the possibly optimizations we could do:

- Create the masks only once, when they are first needed, and [then save the corresponding image](https://developer.mozilla.org/en-US/docs/Web/API/HTMLCanvasElement/toBlob).
- The `masks` canvas does not need to be the full size of the `drawn` canvas; in fact, it only needs to be one pixel in size. The sample code for the book actually does this one. Drawing outside of the canvas is permitted but ignored so it makes things very fast.
- We can also skip drawing sprites that do not intersect with the clicked pixel. We do all of this because our sprites have complicated shapes, but these shapes are always contained in a rectangle, so if the pixel is not in the rectangle of the object, we can safely skip it.

Does any of that make sense? I feel like I'm rambling a bit at this point. Apologies for the wall of text.

Reply:
👍